src/libpriv/rpmostree-core.c

```diff
@@ -2854,8 +2870,37 @@ delete_package_from_root (RpmOstreeContext *self,
       if (!g_str_has_prefix (fn, "usr/"))
         continue;
 
-      if (!glnx_shutil_rm_rf_at (rootfs_dfd, fn, cancellable, error))
-        return FALSE;
+      /* Delete files first, we'll handle directories next */
+      if (S_ISDIR (mode))
+        g_sequence_insert_sorted (dirs_to_remove, g_strdup (fn), compare_strlen, NULL);
+      else
+        {
+          if (unlinkat (rootfs_dfd, fn, 0) < 0)
+            {
+              if (errno != ENOENT)
+                return glnx_throw_errno_prefix (error, "unlinkat(%s)", fn);
+            }
+        }
+    }
+
+  /* We want to perform a bottom-up removal of directories. This is implemented
+   * by having a sequence of filenames sorted by decreasing length.  A filename
+   * of greater length must be a subdirectory, or unrelated.
+   *
+   * If a directory isn't empty, that's fine - a package may own
+   * the directory, but it's valid for other packages to put content under it.
```

Comment:
I think it'll be more correct here actually to gather the seq of directories for all packages, and then do this pass after we've done the files for each pkg.

Reply:
I think that should be fine since we're deleting in order, right?